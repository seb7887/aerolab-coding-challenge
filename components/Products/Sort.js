import React from 'react';

const Sort = ({ total }) => (
  <div>
    <div>
      <span>{total}</span> PRODUCTS
    </div>
  </div>
);

export default Sort;
