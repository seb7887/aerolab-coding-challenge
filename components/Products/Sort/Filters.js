import React from 'react';
import Link from 'next/link';
import styled from 'styled-components';

const Container = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;

  .title {
    margin-right: 1rem;
    @media (max-width: 650px) {
      margin: 0;
      display: none;
    }
  }

  .buttons {
    display: flex;
    flex-direction: row;
    align-items: center;
  }

  a {
    margin-right: 1rem;
    letter-spacing: 0.5px;
    text-align: center;
    padding: 9px 12px;
    border-radius: 20px;
    display: flex;
    justify-content: center;
    align-items: center;
    transition: all 0.3s;
    background-color: ${props => props.theme.bg};
    cursor: pointer;

    &.current {
      background-color: ${props => props.theme.orange};
      color: ${props => props.theme.white};
      cursor: default;
    }

    &:hover {
      background-color: ${props => props.theme.orange};
      color: ${props => props.theme.white};
    }
  }
`;

const Filters = ({ category, sort }) => {
  const isRecent = sort === 'recent';
  const isLow = sort === 'low';
  const isHigh = sort === 'high';

  return (
    <Container>
      <div className='title'>Sort by:</div>
      <div className='buttons'>
        <Link
          prefetch
          href={{
            pathname: '/',
            query: {
              category,
              sort: 'recent',
              page: 0
            }
          }}
        >
          <a className={isRecent ? 'current' : ''} data-testid='sort-filter'>
            Most recent
          </a>
        </Link>
        <Link
          prefetch
          href={{
            pathname: '/',
            query: {
              category,
              sort: 'low',
              page: 0
            }
          }}
        >
          <a className={isLow ? 'current' : ''} data-testid='sort-filter'>
            Lower price
          </a>
        </Link>
        <Link
          prefetch
          href={{
            pathname: '/',
            query: {
              category,
              sort: 'high',
              page: 0
            }
          }}
        >
          <a className={isHigh ? 'current' : ''} data-testid='sort-filter'>
            Higher price
          </a>
        </Link>
      </div>
    </Container>
  );
};

export default Filters;
