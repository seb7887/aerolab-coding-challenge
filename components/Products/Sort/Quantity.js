import React from 'react';
import styled from 'styled-components';

const Container = styled.div`
  color: ${props => props.theme.secondary};
  font-weight: bold;
  padding-right: 1rem;
  border-right: 1px solid ${props => props.theme.primary};

  @media (max-width: 770px) {
    display: none;
  }
`;

const Quantity = ({ quantity, total }) => (
  <Container>
    <span>
      {quantity} of {total}
    </span>{' '}
    products
  </Container>
);

export default Quantity;
