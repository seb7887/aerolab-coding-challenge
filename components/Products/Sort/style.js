import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  font-size: 1rem;
  padding-bottom: 0.8rem;
  border-bottom: 1px solid ${props => props.theme.grey};

  .sort {
    padding-left: 1rem;
    font-weight: 700;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    flex-grow: 2;
  }

  @media (max-width: 650px) {
    font-size: 0.8rem;
  }

  @media (max-width: 500px) {
    font-size: 0.5rem;
  }
`;
