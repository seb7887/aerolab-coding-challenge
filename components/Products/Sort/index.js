import React from 'react';

import Quantity from '../../shared/Quantity';
import Filters from './Filters';
import Pagination from '../../shared/Pagination';

import { Container } from './style';

import config from '../../../config';

const Sort = ({ page, sort, category, total }) => {
  const { perPage } = config;
  const quantity = total > perPage ? (parseInt(page, 10) + 1) * perPage : total;

  return (
    <Container>
      <Quantity quantity={quantity} total={total} />
      <div className='sort'>
        <Filters category={category} sort={sort} />
        <Pagination
          category={category}
          sort={sort}
          page={page}
          total={total}
          perPage={perPage}
        />
      </div>
    </Container>
  );
};

export default Sort;
