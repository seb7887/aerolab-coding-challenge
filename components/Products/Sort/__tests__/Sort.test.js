import { render } from 'react-testing-library';

import Sort from '../index';

jest.mock('next/link', () => {
  return ({ children }) => {
    return children;
  };
});

let page = 0;
const category = 'all';
const sort = 'recent';
let total = 32;

describe('Sort', () => {
  it('renders and matches snapshot', () => {
    const { asFragment } = render(
      <Sort category={category} sort={sort} page={page} total={total} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display correct quantity', () => {
    const { getByText } = render(
      <Sort category={category} sort={sort} page={page} total={total} />
    );

    expect(getByText(/16 of 32/i)).toBeTruthy();
  });

  it('should display three sort filter buttons', () => {
    const { getAllByTestId } = render(
      <Sort category={category} sort={sort} page={page} total={total} />
    );

    expect(getAllByTestId('sort-filter').length).toBe(3);
  });

  it('should display only next button if it is the first page', () => {
    const { getByTestId, queryByTestId } = render(
      <Sort category={category} sort={sort} page={page} total={total} />
    );

    expect(queryByTestId('back-button')).toBeNull();
    expect(getByTestId('next-button')).not.toBeNull();
  });

  it('should display both button if it is not the first or the last page', () => {
    page = 1;
    total = 48;
    const { getByTestId } = render(
      <Sort category={category} sort={sort} page={page} total={total} />
    );

    expect(getByTestId('back-button')).not.toBeNull();
    expect(getByTestId('next-button')).not.toBeNull();
  });

  it('should display only next button if it is the first page', () => {
    page = 1;
    total = 32;
    const { getByTestId, queryByTestId } = render(
      <Sort category={category} sort={sort} page={page} total={total} />
    );

    expect(getByTestId('back-button')).not.toBeNull();
    expect(queryByTestId('next-button')).toBeNull();
  });
});
