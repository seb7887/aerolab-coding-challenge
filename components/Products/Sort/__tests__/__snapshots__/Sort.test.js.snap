// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="style__Container-sc-1cigh81-0 MGrfv"
  >
    <div
      class="Quantity__Container-sc-1ye5v54-0 kMaZJe"
    >
      <span>
        16 of 32
      </span>
       products
    </div>
    <div
      class="sort"
    >
      <div
        class="Filters__Container-sc-1w8ug50-0 eIrHFy"
      >
        <div
          class="title"
        >
          Sort by:
        </div>
        <div
          class="buttons"
        >
          <a
            class="current"
            data-testid="sort-filter"
          >
            Most recent
          </a>
          <a
            class=""
            data-testid="sort-filter"
          >
            Lower price
          </a>
          <a
            class=""
            data-testid="sort-filter"
          >
            Higher price
          </a>
        </div>
      </div>
      <div
        class="Pagination__Container-j6vdk2-0 lhMuFC"
      >
        <a
          data-testid="next-button"
        >
          <img
            alt="Next Page"
            src="/static/images/arrow-right.svg"
          />
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
